File: AegisQuant.Interop.Tests/FfiSafetyTests.cs
using FsCheck;
using FsCheck.Xunit;
using Xunit;

namespace AegisQuant.Interop.Tests;

/// <summary>
/// Property-based tests for FFI safety and Dispose pattern correctness.
/// Feature: aegisquant-hybrid, Property 2: FFI Safety - Error Codes Instead of Panics
/// Validates: Requirements 4.4, 4.5, 8.6
/// </summary>
public class FfiSafetyTests
{
    /// <summary>
    /// Property: For any valid StrategyParams and RiskConfig, creating and disposing
    /// an EngineWrapper should not throw exceptions and should properly release resources.
    /// </summary>
    [Property(MaxTest = 100)]
    public Property EngineWrapper_CreateAndDispose_ShouldNotThrow()
    {
        return Prop.ForAll(
            Arb.From<int>().Filter(x => x > 0 && x < 100),  // ShortMaPeriod
            Arb.From<int>().Filter(x => x > 0 && x < 200),  // LongMaPeriod
            (shortMa, longMa) =>
            {
                // Ensure short < long for valid MA strategy
                var actualShort = Math.Min(shortMa, longMa);
                var actualLong = Math.Max(shortMa, longMa);
                if (actualShort == actualLong) actualLong++;

                var parameters = new StrategyParams
                {
                    ShortMaPeriod = actualShort,
                    LongMaPeriod = actualLong,
                    PositionSize = 100.0,
                    StopLossPct = 0.02,
                    TakeProfitPct = 0.05
                };

                var riskConfig = RiskConfig.Default;

                // Create and dispose should work without throwing
                using var engine = new EngineWrapper(parameters, riskConfig);
                return true;
            });
    }

    /// <summary>
    /// Property: For any EngineWrapper, calling Dispose multiple times should be safe
    /// and not throw exceptions (idempotent dispose).
    /// </summary>
    [Property(MaxTest = 100)]
    public Property EngineWrapper_MultipleDispose_ShouldBeSafe()
    {
        return Prop.ForAll(
            Arb.From<int>().Filter(x => x >= 1 && x <= 10),  // Number of dispose calls
            disposeCount =>
            {
                var engine = new EngineWrapper();

                // Multiple dispose calls should be safe
                for (int i = 0; i < disposeCount; i++)
                {
                    engine.Dispose();
                }

                return true;
            });
    }

    /// <summary>
    /// Property: After disposing an EngineWrapper, any operation should throw
    /// ObjectDisposedException.
    /// </summary>
    [Property(MaxTest = 100)]
    public Property EngineWrapper_AfterDispose_OperationsShouldThrow()
    {
        return Prop.ForAll(
            Arb.From<int>().Filter(x => x >= 0 && x < 3),  // Operation selector
            operationIndex =>
            {
                var engine = new EngineWrapper();
                engine.Dispose();

                try
                {
                    switch (operationIndex)
                    {
                        case 0:
                            engine.GetAccountStatus();
                            break;
                        case 1:
                            engine.ProcessTick(new Tick { Timestamp = 0, Price = 100.0, Volume = 1000.0 });
                            break;
                        case 2:
                            _ = engine.Handle;
                            break;
                    }
                    return false; // Should have thrown
                }
                catch (ObjectDisposedException)
                {
                    return true; // Expected exception
                }
            });
    }


    /// <summary>
    /// Property: For any valid Tick data (positive price, non-negative volume),
    /// ProcessTick should not crash the engine.
    /// </summary>
    [Property(MaxTest = 100)]
    public Property ProcessTick_WithValidData_ShouldNotCrash()
    {
        return Prop.ForAll(
            Arb.From<long>(),  // Timestamp
            Arb.From<double>().Filter(x => x > 0 && x < 1_000_000 && !double.IsNaN(x) && !double.IsInfinity(x)),  // Price
            Arb.From<double>().Filter(x => x >= 0 && x < 1_000_000 && !double.IsNaN(x) && !double.IsInfinity(x)),  // Volume
            (timestamp, price, volume) =>
            {
                using var engine = new EngineWrapper();
                var tick = new Tick
                {
                    Timestamp = timestamp,
                    Price = price,
                    Volume = volume
                };

                try
                {
                    engine.ProcessTick(tick);
                    return true;
                }
                catch (EngineException)
                {
                    // Engine exceptions are acceptable (e.g., no data loaded)
                    return true;
                }
            });
    }

    /// <summary>
    /// Property: GetAccountStatus should always return valid data without crashing.
    /// </summary>
    [Property(MaxTest = 100)]
    public Property GetAccountStatus_ShouldReturnValidData()
    {
        return Prop.ForAll(
            Arb.From<int>().Filter(x => x > 0 && x < 50),  // ShortMaPeriod
            shortMa =>
            {
                var parameters = new StrategyParams
                {
                    ShortMaPeriod = shortMa,
                    LongMaPeriod = shortMa + 10,
                    PositionSize = 100.0,
                    StopLossPct = 0.02,
                    TakeProfitPct = 0.05
                };

                using var engine = new EngineWrapper(parameters, RiskConfig.Default);
                var status = engine.GetAccountStatus();

                // Verify returned data is valid (not NaN or Infinity)
                return !double.IsNaN(status.Balance) &&
                       !double.IsNaN(status.Equity) &&
                       !double.IsNaN(status.Available) &&
                       !double.IsNaN(status.TotalPnl) &&
                       !double.IsInfinity(status.Balance) &&
                       !double.IsInfinity(status.Equity);
            });
    }

    /// <summary>
    /// Property: EngineHandle should correctly report IsInvalid after disposal.
    /// </summary>
    [Property(MaxTest = 100)]
    public Property EngineHandle_AfterDispose_ShouldBeInvalid()
    {
        return Prop.ForAll(
            Arb.From<bool>(),  // Dummy parameter to run multiple times
            _ =>
            {
                var engine = new EngineWrapper();
                var handle = engine.Handle;

                // Before dispose, handle should be valid
                bool validBefore = !handle.IsInvalid;

                engine.Dispose();

                // After dispose, handle should be invalid (closed)
                bool invalidAfter = handle.IsClosed;

                return validBefore && invalidAfter;
            });
    }

    /// <summary>
    /// Property: ErrorHandler.CheckResult should throw appropriate exceptions for error codes.
    /// </summary>
    [Property(MaxTest = 100)]
    public Property ErrorHandler_CheckResult_ShouldThrowForErrors()
    {
        var errorCodes = new[]
        {
            ErrorCodes.NullPointer,
            ErrorCodes.InvalidParam,
            ErrorCodes.EngineNotInit,
            ErrorCodes.RiskRejected,
            ErrorCodes.DataLoadFailed,
            ErrorCodes.InvalidData,
            ErrorCodes.InsufficientCapital,
            ErrorCodes.ThrottleExceeded,
            ErrorCodes.PositionLimit,
            ErrorCodes.InternalPanic
        };

        return Prop.ForAll(
            Gen.Elements(errorCodes).ToArbitrary(),
            errorCode =>
            {
                try
                {
                    ErrorHandler.CheckResult(errorCode, "TestOperation");
                    return false; // Should have thrown
                }
                catch (Exception ex)
                {
                    // Verify appropriate exception type
                    return errorCode switch
                    {
                        ErrorCodes.NullPointer => ex is ArgumentNullException,
                        ErrorCodes.InvalidParam => ex is ArgumentException,
                        ErrorCodes.EngineNotInit => ex is InvalidOperationException,
                        ErrorCodes.RiskRejected => ex is RiskRejectedException,
                        ErrorCodes.DataLoadFailed => ex is DataLoadException,
                        ErrorCodes.InvalidData => ex is InvalidDataException,
                        ErrorCodes.InsufficientCapital => ex is InsufficientCapitalException,
                        ErrorCodes.ThrottleExceeded => ex is ThrottleExceededException,
                        ErrorCodes.PositionLimit => ex is PositionLimitException,
                        ErrorCodes.InternalPanic => ex is EngineException,
                        _ => ex is EngineException
                    };
                }
            });
    }

    /// <summary>
    /// Property: ErrorHandler.CheckResult should not throw for success code.
    /// </summary>
    [Property(MaxTest = 100)]
    public Property ErrorHandler_CheckResult_ShouldNotThrowForSuccess()
    {
        return Prop.ForAll(
            Arb.From<string>().Filter(s => !string.IsNullOrEmpty(s)),
            operation =>
            {
                try
                {
                    ErrorHandler.CheckResult(ErrorCodes.Success, operation);
                    return true;
                }
                catch
                {
                    return false;
                }
            });
    }

    /// <summary>
    /// Unit test: Verify struct sizes match expected C layout.
    /// This is a sanity check for FFI compatibility.
    /// </summary>
    [Fact]
    public void StructSizes_ShouldMatchExpectedLayout()
    {
        // Tick: i64 + f64 + f64 = 8 + 8 + 8 = 24 bytes
        Assert.Equal(24, System.Runtime.InteropServices.Marshal.SizeOf<Tick>());

        // AccountStatus: f64 + f64 + f64 + i32 + f64 = 8 + 8 + 8 + 4 + 8 = 36 bytes
        // Note: May have padding, so check >= 36
        Assert.True(System.Runtime.InteropServices.Marshal.SizeOf<AccountStatus>() >= 36);

        // StrategyParams: i32 + i32 + f64 + f64 + f64 = 4 + 4 + 8 + 8 + 8 = 32 bytes
        Assert.True(System.Runtime.InteropServices.Marshal.SizeOf<StrategyParams>() >= 32);

        // RiskConfig: i32 + f64 + f64 + f64 = 4 + 8 + 8 + 8 = 28 bytes (may have padding)
        Assert.True(System.Runtime.InteropServices.Marshal.SizeOf<RiskConfig>() >= 28);
    }
}

File: aegisquant-core/tests/ffi_safety_tests.rs
//! Property-based tests for FFI safety.
//!
//! Feature: aegisquant-hybrid, Property 2: FFI Safety - Error Codes Instead of Panics
//! Validates: Requirements 2.5, 2.6, 8.1, 8.2, 8.3

use proptest::prelude::*;
use aegisquant_core::ffi::*;
use aegisquant_core::types::*;

proptest! {
    #![proptest_config(ProptestConfig::with_cases(100))]

    /// Property 2: Null engine pointer returns error code, not crash
    #[test]
    fn null_engine_returns_error_code_process_tick(
        timestamp in any::<i64>(),
        price in 0.01f64..1_000_000.0,
        volume in 0.0f64..1_000_000.0
    ) {
        let tick = Tick { timestamp, price, volume };
        
        unsafe {
            let result = process_tick(std::ptr::null_mut(), &tick);
            prop_assert_eq!(result, ERR_NULL_POINTER);
        }
    }

    /// Property 2: Null tick pointer returns error code, not crash
    #[test]
    fn null_tick_returns_error_code(
        _dummy in any::<i32>()  // Just to make it a property test
    ) {
        unsafe {
            let engine = init_engine(std::ptr::null(), std::ptr::null());
            prop_assert!(!engine.is_null());
            
            let result = process_tick(engine, std::ptr::null());
            prop_assert_eq!(result, ERR_NULL_POINTER);
            
            free_engine(engine);
        }
    }

    /// Property 2: Invalid price (<=0) returns error code
    #[test]
    fn invalid_price_returns_error_code(
        timestamp in any::<i64>(),
        price in prop_oneof![
            Just(0.0f64),
            Just(-0.01f64),
            -1_000_000.0f64..0.0
        ],
        volume in 0.0f64..1_000_000.0
    ) {
        let tick = Tick { timestamp, price, volume };
        
        unsafe {
            let engine = init_engine(std::ptr::null(), std::ptr::null());
            prop_assert!(!engine.is_null());
            
            let result = process_tick(engine, &tick);
            prop_assert_eq!(result, ERR_INVALID_DATA);
            
            free_engine(engine);
        }
    }

    /// Property 2: Invalid volume (<0) returns error code
    #[test]
    fn invalid_volume_returns_error_code(
        timestamp in any::<i64>(),
        price in 0.01f64..1_000_000.0,
        volume in -1_000_000.0f64..-0.01
    ) {
        let tick = Tick { timestamp, price, volume };
        
        unsafe {
            let engine = init_engine(std::ptr::null(), std::ptr::null());
            prop_assert!(!engine.is_null());
            
            let result = process_tick(engine, &tick);
            prop_assert_eq!(result, ERR_INVALID_DATA);
            
            free_engine(engine);
        }
    }

    /// Property 2: Valid tick returns success
    #[test]
    fn valid_tick_returns_success(
        timestamp in any::<i64>(),
        price in 0.01f64..1_000_000.0,
        volume in 0.0f64..1_000_000.0
    ) {
        let tick = Tick { timestamp, price, volume };
        
        unsafe {
            let engine = init_engine(std::ptr::null(), std::ptr::null());
            prop_assert!(!engine.is_null());
            
            let result = process_tick(engine, &tick);
            prop_assert_eq!(result, ERR_SUCCESS);
            
            free_engine(engine);
        }
    }

    /// Property 2: Null pointers in get_account_status return error code
    #[test]
    fn get_account_status_null_pointers(
        _dummy in any::<i32>()
    ) {
        unsafe {
            // Null engine
            let mut status = AccountStatus::default();
            let result = get_account_status(std::ptr::null_mut(), &mut status);
            prop_assert_eq!(result, ERR_NULL_POINTER);
            
            // Null status
            let engine = init_engine(std::ptr::null(), std::ptr::null());
            prop_assert!(!engine.is_null());
            
            let result = get_account_status(engine, std::ptr::null_mut());
            prop_assert_eq!(result, ERR_NULL_POINTER);
            
            free_engine(engine);
        }
    }

    /// Property 2: get_account_status with valid pointers returns success
    #[test]
    fn get_account_status_valid_returns_success(
        _dummy in any::<i32>()
    ) {
        unsafe {
            let engine = init_engine(std::ptr::null(), std::ptr::null());
            prop_assert!(!engine.is_null());
            
            let mut status = AccountStatus::default();
            let result = get_account_status(engine, &mut status);
            prop_assert_eq!(result, ERR_SUCCESS);
            prop_assert!(status.balance > 0.0);
            
            free_engine(engine);
        }
    }

    /// Property 2: load_data_from_file with null pointers returns error code
    #[test]
    fn load_data_null_pointers(
        _dummy in any::<i32>()
    ) {
        unsafe {
            let mut report = DataQualityReport::default();
            let path = b"test.csv\0".as_ptr() as *const i8;
            
            // Null engine
            let result = load_data_from_file(std::ptr::null_mut(), path, &mut report);
            prop_assert_eq!(result, ERR_NULL_POINTER);
            
            // Null path
            let engine = init_engine(std::ptr::null(), std::ptr::null());
            let result = load_data_from_file(engine, std::ptr::null(), &mut report);
            prop_assert_eq!(result, ERR_NULL_POINTER);
            
            // Null report
            let result = load_data_from_file(engine, path, std::ptr::null_mut());
            prop_assert_eq!(result, ERR_NULL_POINTER);
            
            free_engine(engine);
        }
    }

    /// Property 2: run_backtest with null engine returns error code
    #[test]
    fn run_backtest_null_engine(
        _dummy in any::<i32>()
    ) {
        unsafe {
            let result = run_backtest(std::ptr::null_mut());
            prop_assert_eq!(result, ERR_NULL_POINTER);
        }
    }

    /// Property 2: Engine initialization with various params never crashes
    #[test]
    fn init_engine_never_crashes(
        short_ma in 1i32..100,
        long_ma in 10i32..500,
        position_size in 0.01f64..100_000.0,
        stop_loss in 0.001f64..0.5,
        take_profit in 0.001f64..1.0,
        max_order_rate in 1i32..100,
        max_position in 0.01f64..1_000_000.0,
        max_order_value in 0.01f64..10_000_000.0,
        max_drawdown in 0.01f64..1.0
    ) {
        let params = StrategyParams {
            short_ma_period: short_ma,
            long_ma_period: long_ma,
            position_size,
            stop_loss_pct: stop_loss,
            take_profit_pct: take_profit,
        };
        
        let risk = RiskConfig {
            max_order_rate,
            max_position_size: max_position,
            max_order_value,
            max_drawdown_pct: max_drawdown,
        };
        
        unsafe {
            let engine = init_engine(&params, &risk);
            prop_assert!(!engine.is_null());
            
            // Verify params were stored correctly
            let engine_ref = &*engine;
            prop_assert_eq!(engine_ref.params.short_ma_period, short_ma);
            prop_assert_eq!(engine_ref.params.long_ma_period, long_ma);
            
            free_engine(engine);
        }
    }

    /// Property 2: Multiple init/free cycles don't cause issues
    #[test]
    fn multiple_init_free_cycles(
        cycles in 1usize..10
    ) {
        for _ in 0..cycles {
            unsafe {
                let engine = init_engine(std::ptr::null(), std::ptr::null());
                prop_assert!(!engine.is_null());
                free_engine(engine);
            }
        }
    }
}
